examples/js/plugins/COGParser.js

```diff
@@ -0,0 +1,306 @@
+ *
+ * @param {Source} source The COGSource
+ * @param {Extent} extent The window extent.
+ * @param {number[]} resolution The spatial resolution of the window.
+ * @returns {number[]} The window.
+ */
+function makeWindowFromExtent(source, extent, resolution) {
+    const [oX, oY] = source.origin;
+    const [imageResX, imageResY] = resolution;
+
+    const wnd = [
+        Math.round((extent.west - oX) / imageResX),
+        Math.round((extent.north - oY) / imageResY),
+        Math.round((extent.east - oX) / imageResX),
+        Math.round((extent.south - oY) / imageResY),
+    ];
+
+    const xMin = Math.min(wnd[0], wnd[2]);
+    let xMax = Math.max(wnd[0], wnd[2]);
+    const yMin = Math.min(wnd[1], wnd[3]);
+    let yMax = Math.max(wnd[1], wnd[3]);
+
+    // prevent zero-sized requests
+    if (Math.abs(xMax - xMin) === 0) {
+        xMax += 1;
+    }
+    if (Math.abs(yMax - yMin) === 0) {
+        yMax += 1;
+    }
+
+    return [xMin, yMin, xMax, yMax];
+}
+
+/**
+ * Creates a texture from the pixel buffer(s).
+ *
+ * @param {Source} source The COGSource
+ * @param {Object} buffers The buffers (one buffer per band)
+ * @returns {THREE.DataTexture} The generated texture.
+ */
+function createTexture(source, buffers) {
+    const { width, height } = buffers;
+    const pixelCount = width * height;
+    const targetDataType = source.dataType;
+
+    let format;
+    let channelCount;
+    switch (buffers.length) {
+        case 1:
+        case 2:
+            format = THREE.RGFormat;
+            channelCount = 2;
+            break;
+        default:
+            format = THREE.RGBAFormat;
+            channelCount = 4;
+            break;
+    }
+
+    let texture;
+    switch (targetDataType) {
+        case THREE.UnsignedByteType: {
+            const buf = new Uint8ClampedArray(pixelCount * channelCount);
+            const data = fillBuffer(source, buf, { noData: source.noData }, source.opaqueByte, buffers);
+            texture = new THREE.DataTexture(data, width, height, format, THREE.UnsignedByteType);
+            break;
+        }
+        case THREE.FloatType: {
+            const buf = new Float32Array(pixelCount * channelCount);
+            const data = fillBuffer(source, buf, { noData: source.noData }, source.opaqueFloat, buffers);
+            texture = new THREE.DataTexture(data, width, height, format, THREE.FloatType);
+            break;
+        }
+        default:
+            throw new Error('unsupported data type');
+    }
+
+    return texture;
+}
+
+
+// Important note : a lot of code is duplicated to avoid putting
+// conditional branches inside loops, as this can severely reduce performance.
+// Note: we don't use Number.isNan(x) in the loops as it slows down the loop due to function
+// invocation. Instead, we use x !== x, as a NaN is never equal to itself.
+/* eslint no-self-compare: 0 */
+function fillBuffer(source, buffers, options, opaqueValue, pixelData) {
```

Comment:
I think we could reduce the number of parameters:
- the parameters with different parameters between calls (i.e. `pixelData` and `buffers`) could be passed as first parameters
- the properties of `source` used here (i.e. `defaultNoData` and `defaultTransparent`) could be passed as a unique `options` object.

```suggestion
function fillBuffer(pixelData, buffers, options) {
```

Reply:
Ok but we need to let opaqueValue as parameter (there are 2 possible values), it's ok for you ?